### PR TITLE
Fixed issues around injecting urls

### LINKED ISSUE #2241: Which endpoints should be propagated to ASPNETCORE_URLS env var?
In #2239, I changed the logic to exclude endpoints which either:
* Don't have a "http" or "https" URI scheme
* Have an EnvironmentVariable set

Is this the logic we want to stick with or is there a more robust check we could make?

### LINKED ISSUE #3477: Adding endpoints to a project without a launch profile results in errors in orchestrator
```C#
var builder = DistributedApplication.CreateBuilder(args);

builder.AddProject<Projects.Stress_ApiService>("stress-apiservice", launchProfileName: null)
       .WithHttpEndpoint(port: 10005, name: "http");

builder.Build().Run();
```

The output

```
fail: Aspire.Hosting.Dcp.dcpctrl.ExecutableReconciler[0]
      could not create Endpoint object  {"Executable": {"name":"stress-apiservice-4slf0l8"}, "Reconciliation": 11, "ServiceName": "stress-apiservice", "Workload": "/stress-apiservice-4slf0l8", "error": "service-producer annotation is invalid: missing information about the port to expose the service"}
fail: Aspire.Hosting.Dcp.dcpctrl.ExecutableReconciler[0]
      could not create Endpoint object  {"Executable": {"name":"stress-apiservice-4slf0l8"}, "Reconciliation": 12, "ServiceName": "stress-apiservice", "Workload": "/stress-apiservice-4slf0l8", "error": "service-producer annotation is invalid: missing information about the port to expose the service"}
```

Seems like there's some strange interaction between DCP when there's no launch profile and adding endpoints.

### PR DESCRIPTION
**Pull request by davidfowl**
ASP.NET Core Urls we only being injected when launch profiles. 

- Fixed issues with ASPNETCORE_URLS not working with launch profiles
- Skip ASPNETCORE_URLS for endpoints that are mapped to an environment
- Added tests

Fixes #2241
Fixes #3477
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3492)

### CODE REVIEW

**Review by karolz-ms — APPROVED** [review]

File: src/Aspire.Hosting/ProjectResourceBuilderExtensions.cs [hunk 1]
```diff
@@ -287,4 +253,53 @@ private static bool IsWebProject(ProjectResource projectResource)
...
+                return;
+            }
+
+            var aspnetCoreUrls = new ReferenceExpressionBuilder();
+
+            var processedHttpsPort = false;
+            var first = true;
+
+            static bool IsValidAspNetCoreUrl(EndpointAnnotation e) =>
+                e.UriScheme is "http" or "https" && e.TargetPortEnvironmentVariable is null;
```
Comment (eerhardt) [inline src/Aspire.Hosting/ProjectResourceBuilderExtensions.cs:278]: Why does having a `TargetPortEnvironmentVariable` mean it isn't a valid ASPNETCORE_URLS?

File: src/Aspire.Hosting/ResourceBuilderExtensions.cs [hunk 2]
```diff
@@ -421,6 +421,8 @@ private static void ApplyEndpoints<T>(this IResourceBuilder<T> builder, IResourc
         // Set the environment variable on the resource
         if (env is not null && builder.Resource is IResourceWithEndpoints resourceWithEndpoints and IResourceWithEnvironment)
         {
+            annotation.TargetPortEnvironmentVariable = env;
```
Comment (eerhardt) [inline src/Aspire.Hosting/ResourceBuilderExtensions.cs:424]: It's kind of weird that `WithEndpoint(..., env: "SOME_ENV")` means that `SOME_ENV` gets the "target port" value.

1. Is this assuming too much this deep?
2. Should `env` be renamed?

File: src/Aspire.Hosting/ProjectResourceBuilderExtensions.cs (hunk 1 above)
Comment (davidfowl) [reply src/Aspire.Hosting/ProjectResourceBuilderExtensions.cs:278]: It's super indirect. I didn't want to design a new API explicitly opting you into ASP.NET Core URLs, but that's another option.

File: src/Aspire.Hosting/ResourceBuilderExtensions.cs (hunk 2 above)
Comment (davidfowl) [reply src/Aspire.Hosting/ResourceBuilderExtensions.cs:424]: Are you concerned about the name of the variable or the behavior? I tried to remove this feature completely, but it makes executables listening on a port more verbose than it needs to be.

File: src/Aspire.Hosting/ProjectResourceBuilderExtensions.cs (hunk 1 above)
Comment (eerhardt) [reply src/Aspire.Hosting/ProjectResourceBuilderExtensions.cs:278]: One thing I think may help is to update the docs on WithEndpoint's "env" variable:

> /// <param name="env">An optional name of the environment variable to inject.</param>

It doesn't really describe what is happening here.

File: src/Aspire.Hosting/ResourceBuilderExtensions.cs (hunk 2 above)
Comment (eerhardt) [reply src/Aspire.Hosting/ResourceBuilderExtensions.cs:424]: > Are you concerned about the name of the variable or the behavior?

A little about the name of the variable. A little about the documentation on what it does. I think I understand the behavior.

File: src/Aspire.Hosting/ProjectResourceBuilderExtensions.cs [hunk 3]
```diff
@@ -287,4 +253,53 @@ private static bool IsWebProject(ProjectResource projectResource)
...
+
+                    processedHttpsPort = true;
+                }
+
+                aspnetCoreUrls.Append($"{e.Property(EndpointProperty.Scheme)}://localhost:{e.Property(EndpointProperty.TargetPort)}");
+                first = false;
+            }
+
+            // Combine into a single expression
+            context.EnvironmentVariables["ASPNETCORE_URLS"] = aspnetCoreUrls.Build();
```
Comment (ReubenBond) [inline src/Aspire.Hosting/ProjectResourceBuilderExtensions.cs:302]: If this is empty, is it better to set it or not set it?

**Review by ReubenBond — APPROVED** [review]

File: src/Aspire.Hosting/ResourceBuilderExtensions.cs (hunk 2 above)
Comment (davidfowl) [reply src/Aspire.Hosting/ResourceBuilderExtensions.cs:424]: Updated

File: src/Aspire.Hosting/ProjectResourceBuilderExtensions.cs (hunk 3 above)
Comment (davidfowl) [reply src/Aspire.Hosting/ProjectResourceBuilderExtensions.cs]: Fixed

**Review by eerhardt — APPROVED** [review]

**Comment by JamesNK** [comment]
Is this the bug I hit yesterday?

**Comment by davidfowl** [comment]
Yep!

**Review by JamesNK — APPROVED** [review]
I don't really know this code but it looks ok to me

**Comment by davidfowl** [comment]
@ReubenBond can you back port your orleans change. This needs to go in afterwards.

**Pull request merged**